### PR TITLE
Testing Farm as Git Hub Action - Typescript version :tractor: 

### PR DESCRIPTION
**Pull request by jamacku**
This PR aims to reimplement the current Action in a language that will allow easier extendability and testability than the current composite yaml/python/bash implementation.

I chose TypeScript because it's the language Git Hub uses to write the official Actions. There are a lot of libraries that make writing GitHub Actions easier than in other languages.

This PR is keeping the current behavior/inputs/outputs except:

- ~~input `git_url` - removed - this value can be gathered from the GitHub context~~
- input `debug` - removed - we should use the native version of running Actions in debug mode - https://docs.github.com/en/actions/monitoring-and-troubleshooting-workflows/enabling-debug-logging

Notes from going through the current implementation:
- input `pr_head_sha` - is this useful/needed?
- ~~outputs - are they used outside of the tfaga Action?~~

#### TODO:
- [x] Tests
- [x] Documentation updates, if needed
- [x] ~~Manual testing of all potential scenarios/ configurations~~
- [x] Review

#### Suggestions
- I suggest installing the Codecov GitHub App. It's a great way to track your test coverage - https://github.com/marketplace/codecov

### CODE REVIEW

File: src/action.ts [hunk 1]
```diff
@@ -0,0 +1,271 @@
...
+    debug(`Testing Farm - state: '${tfResult.state}'`);
+    timeout--;
+
+    await setTimeout(interval);
+  } while (timeout > 0);
+
+  if (timeout === 0) {
+    throw new Error(
+      `Testing Farm - timeout reached. The test is still in state: '${tfResult.state}'`
+    );
```
Comment (jamacku) [inline src/action.ts:176]: Also update status to failed

**Comment by zmiklank** [comment]
Action items for zmiklank - look at how `pr_head_sha` is used in current action.

**Comment by zmiklank** [comment]
> Action items for zmiklank - look at how `pr_head_sha` is used in current action.

So it is used only when using this action to test a PR. Only used to obtain sha of the latest commit in PR. If this can be obtained in some more straightforward way then we do not need this.
Please note, that the testing-farm-as-a-github-action needs to know this SHA without need to cloning the tested repo directly in the GitHub action because for most use-cases is sufficient to clone the tested repo on the TF machine.

Also, if we want to remove this input, even though it seems like non-breaking change, I would consider to release it as tfaga v3.


TODO for @zmiklank: to create needed changes in tfaga-wrapper as discussed in person.

**Comment by zmiklank** [comment]
The 'typescript' branch in tfaga-wrapper was created. I also created draft [PR for nginx](https://github.com/sclorg/nginx-container/pull/267) that moves nginx testing to this typescript-written tfaga once this PR is merged to 'typescript' branch.

**Comment by jamacku** [comment]
@zmiklank, @phracek, I think it's ready for review.

File: .github/workflows/codeql-analysis.yml [hunk 2]
```diff
@@ -0,0 +1,42 @@
+name: "CodeQL"
+
+on:
+  push:
+    branches: [ main, typescript ]
+  pull_request:
+    branches: [ main, typescript ]
+  schedule:
```
Comment (phracek) [inline .github/workflows/codeql-analysis.yml:8]: Why this is called by schedule?

File: action.yml [hunk 3]
```diff
@@ -21,7 +21,7 @@ inputs:
     required: false
     default: 'master'
   github_token:
-    description: 'A github token passed from secrets'
+    description: GitHub TOKEN or PAT
```
Comment (phracek) [inline action.yml:24]: TOKEN is understandable. PAT??? What does it mean?

File: action.yml [hunk 4]
```diff
@@ -87,280 +83,17 @@ inputs:
     description: 'HEAD SHA of a Pull Request. Used for posting statuses to the PR. The value is obtained from `git rev-parse HEAD` if this input is not set.'
     required: false
     default: ''
+  timeout:
+    description: 'Action is waiting for testing farm to finish or until timeout is reached.'
+    required: false
+    default: '480'
```
Comment (phracek) [inline action.yml:89]: Please write down, that 480 does it mean.

File: src/main.ts [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+import '@total-typescript/ts-reset';
+
+import action from './action';
+
+const octokit = new Octokit({
+  auth: getInput('github_token', { required: true }),
+});
+
+try {
+  await action(octokit);
```
Comment (phracek) [inline src/main.ts:13]: Can you please add there a comment, where is action called? Even it is file action.ts.

**Review by phracek — CHANGES_REQUESTED** [review]

File: action.yml (hunk 3 above)
Comment (jamacku) [reply action.yml:24]: PAT is an abbreviation of Personal Access Token - https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/managing-your-personal-access-tokens

File: .github/workflows/codeql-analysis.yml (hunk 2 above)
Comment (jamacku) [reply .github/workflows/codeql-analysis.yml:8]: It's from a template provided directly by GitHub/CodeQL. The reason is to have new scan results even if there is no activity on the repository.

It's optional. I can drop it if you don't like it.

File: src/main.ts (hunk 5 above)
Comment (jamacku) [reply src/main.ts:15]: Sure, done

File: action.yml (hunk 4 above)
Comment (jamacku) [reply action.yml:89]: I have added a note that the value is in minutes.

**Comment by jamacku** [comment]
@phracek, I have added a `DEVELOPMENT.md` document. It should help you and others get started with Typescript and GitHub Actions.

File: .github/workflows/codeql-analysis.yml (hunk 2 above)
Comment (phracek) [reply .github/workflows/codeql-analysis.yml:8]: I don't think that this is needed.

**Comment by phracek** [comment]
@jamacku I see there is checkbox called `Manual testing of all potential scenarios/ configurations`. What does it mean?

File: .github/workflows/codeql-analysis.yml (hunk 2 above)
Comment (jamacku) [reply .github/workflows/codeql-analysis.yml]: I have dropped `on: schedule` trigger.

**Comment by jamacku** [comment]
@phracek, It was just a note for me to try to test it "manually" using my fork of nginx-container, but I think the best would be to test it directly in nginx-container upstream. 

I did some basic tests and the basics work.

**Pull request merged**